### PR TITLE
Add word2vec with wmd distance similarity

### PR DESCRIPTION
**Pull request by ayush-1506**
Fixes : #637

To do : 

- [x] Make this faster

- [ ] get_similar_bugs returns the 5 closest bugs. Experiment with minimum distance cut off along with this.

### CODE REVIEW

**Comment by marco-c** [comment]
Can you post some example results?

What part is slow? And how slow?

**Comment by marco-c** [comment]
> Can you post some example results?

I mean the evaluation results.

**Comment by ayush-1506** [comment]
Ok, so the first part is done (implemented r-wmd with cosine distance instead of wmd), So it's comparatively fast now.

**Comment by marco-c** [comment]
Good! Let me know when you are done with an evaluation so we can compare the results with the others.

**Comment by ayush-1506** [comment]
Looks like there's an edge case that I'm overlooking. This causes an error in a few specific bugs. I'll try to narrow it down.

**Comment by ayush-1506** [comment]
@marco-c 
```
Recall: 24.818577648766325%
Precision: 3.189405949827474%
```

**Comment by marco-c** [comment]
Are these recall and precision with the same number of bugs you used for the other evaluations? If not, can you run at least the nn evaluation with the same number of bugs so we can compare?

**Comment by marco-c** [comment]
The next step will be #631, we need to understand why the results are like this and if/how we can improve them.

**Comment by ayush-1506** [comment]
I tried nn similarity with the same number of bugs:
```
Recall: 30.327056491575817%
Precision: 3.912793299661147%
```

File: bugbug/similarity.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 import re
 from collections import defaultdict
 
+import numpy as np
+from gensim.models import Word2Vec
```
Comment (marco-c) [inline bugbug/similarity.py]: You'll need to move the import where the other gensim imports are

File: bugbug/similarity.py [hunk 2]
```diff
@@ -88,6 +90,7 @@ def evaluation(self):
         hits_p = 0
 
         for bug in bugzilla.get_bugs():
+
```
Comment (marco-c) [inline bugbug/similarity.py]: Please remove this newline as it's unrelated to your changes

File: bugbug/similarity.py [hunk 3]
```diff
@@ -175,3 +178,46 @@ def get_similar_bugs(self, query):
...
+
+class Word2VecWmdSimilarity(BaseSimilarity):
+    def __init__(self):
+        self.corpus = []
+        self.bug_ids = []
+        for bug in bugzilla.get_bugs():
+            self.corpus.append(text_preprocess(get_text(bug)))
+            self.bug_ids.append(bug["id"])
+        self.w2vmodel = Word2Vec(self.corpus, size=100, min_count=1)
```
Comment (marco-c) [inline bugbug/similarity.py]: Maybe better to use min_count=5?

Comment (marco-c) [reply bugbug/similarity.py]: That is, the default

File: bugbug/similarity.py [hunk 4]
```diff
@@ -175,3 +178,46 @@ def get_similar_bugs(self, query):
...
+
+    def get_similar_bugs(self, query):
+        self.w2vmodel.init_sims()
```
Comment (marco-c) [inline bugbug/similarity.py]: Maybe we can call this in the constructor?

File: bugbug/similarity.py (hunk 3 above)
Comment (marco-c) [inline bugbug/similarity.py]: Maybe we should shuffle the corpus

File: bugbug/similarity.py (hunk 4 above)
Comment (marco-c) [reply bugbug/similarity.py]: And you can use replace=True to reduce the memory usage

File: bugbug/similarity.py [hunk 5]
```diff
@@ -175,3 +178,46 @@ def get_similar_bugs(self, query):
...
+                self.w2vmodel.wv.vectors_norm,
+                self.w2vmodel.wv.vectors_norm[
+                    [self.w2vmodel.wv.vocab[word].index for word in words]
+                ].transpose(),
+            ),
+            dtype=np.double,
+        )
+        distances = []
+        for i in range(len(self.corpus)):
+            idx = [self.w2vmodel.wv.vocab[word].index for word in self.corpus[i]]
```
Comment (marco-c) [inline bugbug/similarity.py]: Nit: it's plural, so maybe use `indexes`

File: bugbug/similarity.py [hunk 6]
```diff
@@ -175,3 +178,46 @@ def get_similar_bugs(self, query):
...
+                word_dists = all_distances[idx]
+                rwmd = max(
+                    np.sum(np.min(word_dists, axis=0)),
+                    np.sum(np.min(word_dists, axis=1)),
+                )
+
+                distances.append((self.bug_ids[i], rwmd))
+
+        distances.sort(key=lambda v: v[1])
+
```
Comment (marco-c) [inline bugbug/similarity.py]: You should now apply real WMD, like at https://github.com/marco-c/crashsimilarity/blob/master/crashsimilarity/models/base.py#L165.

File: bugbug/similarity.py [hunk 7]
```diff
@@ -175,3 +180,111 @@ def get_similar_bugs(self, query):
...
+    def __init__(self):
+        self.corpus = []
+        self.bug_ids = []
+        for bug in bugzilla.get_bugs():
+            self.corpus.append(text_preprocess(get_text(bug)))
+            self.bug_ids.append(bug["id"])
+        self.w2vmodel = Word2Vec(self.corpus, size=100, min_count=5)
+        self.w2vmodel.init_sims(replace=True)
+
+    def wmdistance(self, document1, document2, all_distances, distance_metric="cosine"):
```
Comment (marco-c) [inline bugbug/similarity.py]: Can you check the differences between this function and the equivalent present in gensim? I don't remember exactly why I rewrote it.

**Comment by ayush-1506** [comment]
Results with the latest implementation:
```
Recall: 25.82417582417583%
Precision: 3.0921052631578947%
```

File: bugbug/similarity.py (hunk 7 above)
Comment (ayush-1506) [reply bugbug/similarity.py]: There is a subtle difference. Here we've computed the cosine distances before calling wmdistance and stored them in all_distances. (and then we use this inside the wmdistance method.). On the other hand the word2vec [wmdistance](https://tedboy.github.io/nlps/_modules/gensim/models/word2vec.html#Word2Vec.wmdistance) calculates them inside the method. 

The other major difference is that `word2vec.wmdistance` calculates only the euclidean distance. (There isn't an option to calculate cosine distances.), so we need to implement it separately over here.

Comment (marco-c) [reply bugbug/similarity.py]: Can you add a comment explaining that the function partially comes from gensim and explaining the differences?

**Pull request merged**